### PR TITLE
ROX-19051: add ScannerV4 to SecuredCluster CRD

### PR DESCRIPTION
**Pull request by johannes94**
## Description

Add Scanner V4 to the SecuredCluster CRD, and enable operator to manage the local Scanner V4 components according to the CRD.

## Checklist
- [x] Investigated and inspected CI test results
- [x] Unit test and regression tests added
- [x] Evaluated and added CHANGELOG entry if required
- [x] Determined and documented upgrade steps
- [x] Documented user facing changes (create PR based on [openshift/openshift-docs](https://github.com/openshift/openshift-docs) and merge into [rhacs-docs](https://github.com/openshift/openshift-docs/tree/rhacs-docs))

If any of these don't apply, please comment below.

## Testing Performed

**Test central and secured cluster in different cluster**

- Running a local build of operator against an OCP 4.14 cluster.
- Create a central CR from `./operator/tests/common/central-cr.yaml`
- Wait for resources to become ready
- Create an init bundle and apply to the namespace
- Create a securedcluster CR from `./operator/tests/common/secured-cluster-cr.yaml`
- Verify Scanner V4 is not deployed, but it's env variables for sensor are set

**Test central and secured cluster in different namespaces(via external route)**

- Running a local build of operator against an OCP 4.14 cluster
- Create a central CR in namespace `stackrox` from `./operator/tests/common/central-cr.yaml`
- Edit the CR so that it exposes UI other route
- Wait for resources to become ready
- Create init bundle and apply to namespace `secured-cluster`
- Add centralEndpoint and create a securedcluster CR from `./operator/tests/common/secured-cluster-cr.yaml`
- Scanner V4 resources should be created in secured-cluster namespace as if it was a external cluster
- disable Scanner V4 for secured cluster
- Verify that Scanner V4 resources get removed

### Reminder for reviewers

In addition to reviewing code here, reviewers **must** also review testing and request further testing in case the
performed one does not seem sufficient. As a reviewer, you must not approve the change until you understand the
performed testing and you are satisfied with it.

### CODE REVIEW

**Review by RTann — COMMENTED** [review]
assuming all YAML files are generated? going to ignore them

File: operator/pkg/securedcluster/values/translation/base-values.yaml [hunk 1]
```diff
@@ -10,3 +10,15 @@ meta:
   useLookup: false
 allowNonstandardReleaseName: true
 allowNonstandardNamespace: true
+scannerV4:
```
Comment (RTann) [inline operator/pkg/securedcluster/values/translation/base-values.yaml:13]: how come we didn't need this for Scanner v2?

File: operator/pkg/securedcluster/extensions/reconcile_scanner_v4_db_password.go [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+func reconcileScannerV4DBPassword(ctx context.Context, s *platform.SecuredCluster, client ctrlClient.Client, _ func(updateStatusFunc), _ logr.Logger) error {
+	config, err := scanner.AutoSenseLocalScannerV4Config(ctx, client, *s)
+	if err != nil {
+		return err
+	}
+
+	// Only reconcile password if resources are deployed with the SecuredCluster.
+	securedClusterWithScannerV4 := &securedClusterWithScannerV4Bearer{
+		SecuredCluster:   s,
+		scannerV4Enabled: config.DeployScannerResources,
```
Comment (RTann) [inline operator/pkg/securedcluster/extensions/reconcile_scanner_v4_db_password.go:41]: just confirming there isn't a ScannerV4 equivalent

File: operator/apis/platform/v1alpha1/securedcluster_types.go [hunk 3]
```diff
@@ -340,6 +344,24 @@ type LocalScannerComponentSpec struct {
 	DB *DeploymentSpec `json:"db,omitempty"`
 }
 
+// LocalScannerV4ComponentSpec defines settings for the "scanner V4" component in SecuredClusters
+type LocalScannerV4ComponentSpec struct {
+	// If you do not want to deploy the Red Hat Advanced Cluster Security local Scanner V4, you can disable it here
+	// (not recommended).
```
Comment (mclasmeier) [inline operator/apis/platform/v1alpha1/securedcluster_types.go:350]: Maybe mention here again the relationship with `scanner.disable`? Not sure myself.

File: operator/apis/platform/v1alpha1/securedcluster_types.go [hunk 4]
```diff
@@ -340,6 +344,24 @@ type LocalScannerComponentSpec struct {
...
+	// If you do so, all the settings in this section will have no effect.
+	//+kubebuilder:default=AutoSense
+	//+operator-sdk:csv:customresourcedefinitions:type=spec,displayName="Scanner Component",order=1
+	ScannerComponent *LocalScannerComponentPolicy `json:"scannerComponent,omitempty"`
+
+	// Settings pertaining to the indexer deployment.
+	//+operator-sdk:csv:customresourcedefinitions:type=spec,order=2,xDescriptors={"urn:alm:descriptor:com.tectonic.ui:fieldDependency:.scannerComponent:AutoSense"}
```
Comment (mclasmeier) [inline operator/apis/platform/v1alpha1/securedcluster_types.go:357]: What does `fieldDependency:.scannerComponent:AutoSense` mean here?
I assume `scannerComponent` and also be set to something like `Enabled` -- would this have impact on this `Indexer` field?

File: operator/apis/platform/v1alpha1/securedcluster_types.go [hunk 5]
```diff
@@ -340,6 +344,24 @@ type LocalScannerComponentSpec struct {
...
+	Indexer *ScannerV4Component `json:"indexer,omitempty"`
+
+	// Settings pertaining to the DB deployment.
+	//+operator-sdk:csv:customresourcedefinitions:type=spec,order=3,xDescriptors={"urn:alm:descriptor:com.tectonic.ui:fieldDependency:.scannerComponent:AutoSense"}
```
Comment (mclasmeier) [inline operator/apis/platform/v1alpha1/securedcluster_types.go:361]: Same question as above here.

File: operator/pkg/securedcluster/values/translation/base-values.yaml (hunk 1 above)
Comment (mclasmeier) [reply operator/pkg/securedcluster/values/translation/base-values.yaml:13]: But isn't it here
https://github.com/stackrox/stackrox/blob/14ec4c7db56d9e1248ad0e4bfd10400b6934793f/operator/pkg/central/values/translation/base-values.yaml#L34?

File: operator/pkg/securedcluster/values/translation/image_overrides.go [hunk 6]
```diff
@@ -4,10 +4,13 @@ import "github.com/stackrox/rox/operator/pkg/images"
...
-		images.ScannerSlim:   "image.scanner.fullRef",
-		images.ScannerSlimDB: "image.scannerDb.fullRef",
+		images.Main:             "image.main.fullRef",
+		images.CollectorSlim:    "image.collector.slim.fullRef",
+		images.CollectorFull:    "image.collector.full.fullRef",
+		images.ScannerSlim:      "image.scanner.fullRef",
+		images.ScannerSlimDB:    "image.scannerDb.fullRef",
+		images.ScannerV4DB:      "image.scannerV4.db.fullRef",
+		images.ScannerV4Indexer: "scannerV4.indexer.image.fullRef",
+		images.ScannerV4Matcher: "scannerV4.matcher.image.fullRef",
```
[the remark this answers is not in the document]
Comment (mclasmeier) [reply operator/pkg/securedcluster/values/translation/image_overrides.go:14]: There is definitely a lot of opportunity for confusion here.

tl;dr: I believe the values above should be in-line with the rest of the images listed, that is `image.scannerV4.fullRef` and `image.scannerV4DB.fullRef`.

Reason: See https://github.com/stackrox/stackrox/blob/14ec4c7db56d9e1248ad0e4bfd10400b6934793f/image/templates/helm/stackrox-secured-cluster/templates/_init.tpl.htpl#L292.

Internally, in the (shared) scanner V4 init code we expect the image configuration in `scannerV4.<component>.image`.

But to not break the consistency within the secured-cluster chart we expect the configurations be reside in `image.scannerV4` and `image.scannerV4DB` -- this is simply how it is being done there for all the other images.

Then we do this here: https://github.com/stackrox/stackrox/blob/14ec4c7db56d9e1248ad0e4bfd10400b6934793f/image/templates/helm/stackrox-secured-cluster/templates/_init.tpl.htpl#L297
which copies the input we get from the user of the Helm chart (from `image.scannerV4`) and copy it to the path `scannerV4....`.
So that the image configuration code here https://github.com/stackrox/stackrox/blob/14ec4c7db56d9e1248ad0e4bfd10400b6934793f/image/templates/helm/shared/templates/_scanner-v4_init.tpl.htpl#L81 can work with it.

Since here, in this case, the operator can be regarded as the user of the Helm chart, I believe it should be what I wrote in the tl;dr above.

**Review by mclasmeier — CHANGES_REQUESTED** [review]
Not done yet.

**Review by mclasmeier — CHANGES_REQUESTED** [review]

File: operator/apis/platform/v1alpha1/securedcluster_types.go (hunk 4 above)
Comment (johannes94) [reply operator/apis/platform/v1alpha1/securedcluster_types.go:357]: For LocalScannerComponentPolicy the enum has two possible values `AutoSense` or `Disabled`. As far as I understood this dependency hides the indexer config fields if the value for ScannerComponent is not AutoSense

File: operator/config/manifests/bases/rhacs-operator.clusterserviceversion.yaml [hunk 7]
```diff
@@ -752,6 +752,29 @@ spec:
...
         displayName: Scanner Component
         path: scanner.scannerComponent
+      - displayName: Persistence
+        path: scannerV4.db.persistence
+      - description: Uses a Kubernetes persistent volume claim (PVC) to manage the
+          storage location of persistent data. Recommended for most users.
+        displayName: Persistent volume claim
+        path: scannerV4.db.persistence.persistentVolumeClaim
+      - description: The name of the PVC to manage persistent data. If no PVC with
+          the given name exists, it will be created. Defaults to "stackrox-db" if
```
[the remark this answers is not in the document]
Comment (johannes94) [reply operator/config/manifests/bases/rhacs-operator.clusterserviceversion.yaml:762]: It should be, and it is because PersistentVolumeClaim is reused. The comment is true there's a kubebuilder annotation that is setting this `stackrox-db` as default claim name.

I wonder why the volume creation still works with the correct name if nothing is configured. Give me some time to check that.

File: operator/config/manifests/bases/rhacs-operator.clusterserviceversion.yaml [hunk 8]
```diff
@@ -752,6 +752,29 @@ spec:
...
+          not set.
+        displayName: Claim Name
+        path: scannerV4.db.persistence.persistentVolumeClaim.claimName
+      - description: Controls the number of replicas and autoscaling for this component.
+        displayName: Scaling
+        path: scannerV4.indexer.scaling
+      - description: When enabled, the number of analyzer replicas is managed dynamically
```
[the remark this answers is not in the document]
Comment (johannes94) [reply operator/config/manifests/bases/rhacs-operator.clusterserviceversion.yaml]: changed it to component, this is also an effect of a shared type

File: operator/pkg/securedcluster/extensions/reconcile_scanner_v4_db_password.go (hunk 2 above)
Comment (johannes94) [reply operator/pkg/securedcluster/extensions/reconcile_scanner_v4_db_password.go:41]: Correct, we're using the AutoSense struct for both scanner V2 and V4

File: operator/pkg/securedcluster/values/translation/translation_test.go [hunk 9]
```diff
@@ -131,6 +131,9 @@ func (s *TranslationTestSuite) TestTranslate() {
 				"scanner": map[string]interface{}{
 					"disable": false,
 				},
+				"scannerV4": map[string]interface{}{
+					"disable": false,
```
[the remark this answers is not in the document]
Comment (johannes94) [reply operator/pkg/securedcluster/values/translation/translation_test.go:135]: I wasn't aware that it was supposed to turned off by default. I did this because also old scanner was enabled in this test.

File: operator/pkg/securedcluster/scanner/auto_sense.go [hunk 10]
```diff
@@ -35,9 +35,24 @@ func AutoSenseLocalScannerConfig(ctx context.Context, client ctrlClient.Client,
 	SetScannerDefaults(&s.Spec)
 	scannerComponent := *s.Spec.Scanner.ScannerComponent
 
+	return autoSenseScanner(ctx, client, scannerComponent, s.Namespace)
+}
+
+// AutoSenseLocalScannerV4Config detects whether the local scanner V4 should be deployed and/or used by sensor.
+// Takes into account the setting in provided SecuredCluster CR as well as the presence of a Central instance in the same namespace.
+// Modifies the provided SecuredCluster object to set a default Spec.ScannerV4 if missing.
+func AutoSenseLocalScannerV4Config(ctx context.Context, client ctrlClient.Client, s platform.SecuredCluster) (AutoSenseResult, error) {
```
[the remark this answers is not in the document]
Comment (johannes94) [reply operator/pkg/securedcluster/scanner/auto_sense.go:44]: I agree the logic was in place before we started with Scanner V4 work. So apparently we had following behaviour:
- helm chart install on openshift/non-openshift -> disable local scanner by default
- operator install on openshift -> enable local scanner by default
- operator install on non openshift (possible, but not officially supported) -> disable local scanner by default
@mclasmeier Is this correct?

Not sure if we should change the helm defaults, in K8s changes of default values are considered breaking changes.

**Review by johannes94 — COMMENTED** [review]
Addressed some comments. need to look more into some others to get them resolved.

File: operator/pkg/securedcluster/values/translation/image_overrides.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (johannes94) [reply operator/pkg/securedcluster/values/translation/image_overrides.go]: You're right. Thanks for the context Moritz I've changed the image reference and double checked they are correctly passed through.

File: operator/config/manifests/bases/rhacs-operator.clusterserviceversion.yaml (hunk 7 above)
[the remark this answers is not in the document]
Comment (johannes94) [reply operator/config/manifests/bases/rhacs-operator.clusterserviceversion.yaml]: So apparently K8s does only apply the defaults if the parent object in the CR is not nil, since in my tests I did not set any values for persistence, the default name defined here was not applied.

In order to support correct default naming if persistenceVolumeClaim is not nil I had to introduce a dedicated ScannerV4Persistence/ScannerV4PersistentVolumeClaim struct.

File: operator/apis/platform/v1alpha1/common_types.go [hunk 11]
```diff
@@ -5,6 +5,7 @@ package v1alpha1
 import (
 	corev1 "k8s.io/api/core/v1"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/utils/pointer"
```
Comment (mclasmeier) [inline operator/apis/platform/v1alpha1/common_types.go:8]: According to https://pkg.go.dev/k8s.io/utils/pointer it is deprecated.

File: operator/pkg/securedcluster/values/translation/image_overrides.go [hunk 12]
```diff
@@ -4,10 +4,13 @@ import "github.com/stackrox/rox/operator/pkg/images"
...
-		images.Main:          "image.main.fullRef",
-		images.CollectorSlim: "image.collector.slim.fullRef",
-		images.CollectorFull: "image.collector.full.fullRef",
-		images.ScannerSlim:   "image.scanner.fullRef",
-		images.ScannerSlimDB: "image.scannerDb.fullRef",
+		images.Main:             "image.main.fullRef",
+		images.CollectorSlim:    "image.collector.slim.fullRef",
+		images.CollectorFull:    "image.collector.full.fullRef",
+		images.ScannerSlim:      "image.scanner.fullRef",
+		images.ScannerSlimDB:    "image.scannerDb.fullRef",
```
[the remark this answers is not in the document]
Comment (mclasmeier) [reply operator/pkg/securedcluster/values/translation/image_overrides.go:11]: I understood @RTann's questions to refer to the Scanner v2 images in l10 and l11.

**Comment by johannes94** [comment]
Had to rebase, @RTann @mclasmeier use https://github.com/stackrox/stackrox/pull/9375/files/1c2bad201d1ba6000ce193c239ca7ae781ccdd16..29e3c26bc83a139ab3180d11f52e3c20f74f9298 to see changes since you last review.

File: operator/pkg/securedcluster/extensions/reconcile_scanner_v4_db_password.go [hunk 13]
```diff
@@ -0,0 +1,45 @@
...
+
+	"github.com/go-logr/logr"
+	"github.com/operator-framework/helm-operator-plugins/pkg/extensions"
+	platform "github.com/stackrox/rox/operator/apis/platform/v1alpha1"
+	commonExtensions "github.com/stackrox/rox/operator/pkg/common/extensions"
+	"github.com/stackrox/rox/operator/pkg/securedcluster/scanner"
+	ctrlClient "sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+var _ commonExtensions.ScannerBearingCustomResource = (*securedClusterWithScannerBearer)(nil)
```
Comment (RTann) [inline operator/pkg/securedcluster/extensions/reconcile_scanner_v4_db_password.go:14]: just noticed this should be `securedClusterWithScannerV4Bearer`. Sorry this whole V4 thing is being a huge pain

File: operator/apis/platform/v1alpha1/common_types.go [hunk 14]
```diff
@@ -146,13 +147,67 @@ type ScannerV4Component struct {
...
+		return ""
+	}
+	if p.HostPath == nil {
+		return ""
+	}
+
+	return pointer.StringDeref(p.HostPath.Path, "")
+}
+
+// ScannerV4PersistentVolumeClaim defines PVC-based persistence settings for Central DB.
```
[the remark this answers is not in the document]
Comment (johannes94) [reply operator/apis/platform/v1alpha1/common_types.go]: Done.

File: operator/pkg/securedcluster/extensions/reconcile_scanner_v4_db_password.go (hunk 13 above)
Comment (johannes94) [reply operator/pkg/securedcluster/extensions/reconcile_scanner_v4_db_password.go]: Done.

File: operator/apis/platform/v1alpha1/common_types.go (hunk 11 above)
Comment (mclasmeier) [reply operator/apis/platform/v1alpha1/common_types.go:8]: Ok, how about creating a separate ticket for getting rid of deprecated dependencies such as these?

File: operator/pkg/securedcluster/values/translation/base-values.yaml (hunk 1 above)
Comment (mclasmeier) [reply operator/pkg/securedcluster/values/translation/base-values.yaml:13]: I'd suggest to track these questions with a ticket. WDYT?

File: operator/pkg/securedcluster/values/translation/translation.go [hunk 15]
```diff
@@ -98,7 +99,12 @@ func (t Translator) translate(ctx context.Context, sc platform.SecuredCluster) (
 		return nil, err
 	}
 
-	v.AddChild("sensor", t.getSensorValues(sc.Spec.Sensor, scannerAutoSenseConfig))
+	scannerV4AutoSenseConfig, err := scanner.AutoSenseLocalScannerV4Config(ctx, t.client, sc)
```
Comment (mclasmeier) [inline operator/pkg/securedcluster/values/translation/translation.go:102]: Just thinking loud here, feel free to resolve: Should we handle errors in a situation like this more forgiving, for example by picking a default value and continuing with the translation machinery?

File: operator/pkg/securedcluster/values/translation/translation.go [hunk 16]
```diff
@@ -113,6 +119,9 @@ func (t Translator) translate(ctx context.Context, sc platform.SecuredCluster) (
 	}
 
 	v.AddChild("scanner", t.getLocalScannerComponentValues(sc, scannerAutoSenseConfig))
+	if sc.Spec.ScannerV4 != nil && features.ScannerV4Support.Enabled() {
```
Comment (mclasmeier) [inline operator/pkg/securedcluster/values/translation/translation.go:122]: General remark: I fear we need to discuss a little bit more about the exact semantics of scanner (v4) activation and defaulting. Background: it is not clear to me (yet) how the operator-way and the Helm chart (which also includes logic for figuring out which components to deploy) interact or overlap.

We can do this on a different ticket.

Comment (mclasmeier) [reply operator/pkg/securedcluster/values/translation/translation.go:122]: Feel free to resolve.

**Review by mclasmeier — APPROVED** [review]
Just some minor remarks.

File: operator/pkg/securedcluster/values/translation/translation.go (hunk 15 above)
Comment (johannes94) [reply operator/pkg/securedcluster/values/translation/translation.go:102]: Good though, may be worth. Since I'm not aware of any issues with the implementation right now I think we should leave it as is for this PR and reassess at another point in time.

File: operator/pkg/securedcluster/values/translation/translation.go (hunk 16 above)
Comment (johannes94) [reply operator/pkg/securedcluster/values/translation/translation.go:122]: Matches my gut feeling, let's discuss this and open another PR if we'd like a change.

File: operator/apis/platform/v1alpha1/securedcluster_types.go (hunk 3 above)
Comment (johannes94) [reply operator/apis/platform/v1alpha1/securedcluster_types.go:360]: See above.

File: operator/apis/platform/v1alpha1/securedcluster_types.go [hunk 17]
```diff
@@ -68,16 +68,20 @@ type SecuredClusterSpec struct {
 	//+operator-sdk:csv:customresourcedefinitions:type=spec,order=7,displayName="Scanner Component Settings"
 	Scanner *LocalScannerComponentSpec `json:"scanner,omitempty"`
 
+	// Settings for the Scanner V4 component, which can run in addition to the existing Scanner service
+	//+operator-sdk:csv:customresourcedefinitions:type=spec,order=8,displayName="Scanner V4 Component Settings"
```
[the remark this answers is not in the document]
Comment (johannes94) [reply operator/apis/platform/v1alpha1/securedcluster_types.go:72]: No I understand,  as suggested let's discuss defaults for helm / operator in general and adjust as needed after the merge.
